storage: Don't GC a replica whose ID has been incremented

a-robinson opened the pull request:
If the ID has changed since replica GC started its processing, then it
has changed its identity within the raft group and might not merit
removal anymore.

Could potentially be worked into a replacement for #14344, but I think it's safer to consider as a supplement given the non-contiguous locking involved. If this looks good to you as a supplement, I'll have to change the new test in #14344 since it was relying on this check not working.

cc #14231 @bdarnell @tamird

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/14348)
<!-- Reviewable:end -->

a-robinson committed:
storage: Don't GC a replica whose ID has been incremented

If the ID has changed since replica GC started its processing, then it
has changed its identity within the raft group and might not merit
removal anymore.

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/storage/store.go, line 2215 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14348#-KfwnMM1mY-p8MpqDbSq:-KfwnMM1mY-p8MpqDbSr:bvtw2sc) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae937f54a7cd173124e6e11fc4b7ab46dd6fba50/pkg/storage/store.go#L2215)):*
> ```Go
> 
> 	rep.mu.Lock()
> 	if rep.mu.replicaID >= consistentDesc.NextReplicaID {
> ```

Add a comment that we check both rep.mu.replicaID and rep.mu.desc because they sometimes differ (and link to the issue).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14348#-:-KfwnTKX62m-xTqlRpPK:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tamird commented:
So we're continuing down the road of allowing replica IDs to change? I thought there was @bdarnell's comment on the other PR indicated we'd like to rethink this.

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14348#-:-KfwnmaPpAFvwioDB6ZJ:b-4pd5t3)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
We can rethink it, but that doesn't have to happen *right now*, and in the meantime we should make it safe.

a-robinson commented:
Opened https://github.com/cockroachdb/cockroach/issues/14349 to tracker the bigger change

---

Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/storage/store.go, line 2215 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14348#-KfwnMM1mY-p8MpqDbSq:-KfwyhHSCO_-FoPP7tCW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae937f54a7cd173124e6e11fc4b7ab46dd6fba50/pkg/storage/store.go#L2215)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Add a comment that we check both rep.mu.replicaID and rep.mu.desc because they sometimes differ (and link to the issue).
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14348#-:-KfwzY0j_Id0ae_cQDrp:b-65l3xq)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
Rebased on top of #14344 and modified the test as needed to reliably trigger the race.

---

Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14348#-:-Kfx4Qiskxqjh264SlF0:b-6gduae)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14348)*
<!-- Sent from Reviewable.io -->
